modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
+  // as well we will get a duplicate symbol error. I believe this is due to
+  // module load ordering issues.
+  // TODO: make a proc in ChapelLocale that returns this rather than using the extern
+  extern var chpl_nodeID: chpl_nodeID_t;
+
+  // TODO: hook into chpl_here_alloc and friends somehow
+  // The runtime exits on errors for these, no need to check for nil on return
+  pragma "insert line file info"
+  extern proc chpl_mem_alloc(size: size_t,
+                             description: chpl_mem_descInt_t): c_void_ptr;
 
-  pragma "default string value" extern var defaultStringValue: string = noinit;
+  pragma "insert line file info"
+  extern proc chpl_mem_calloc(number: size_t, size: size_t,
+                              description: chpl_mem_descInt_t) : c_void_ptr;
 
-  // The following method is called by the compiler to determine the default
-  // value of a given type.  For strings, this should return the value
-  // defined above.
-  // strings will be records, so param is not possible for them
-  inline proc _defaultOf(type t) where t: string return defaultStringValue;
+  pragma "insert line file info"
+  extern proc chpl_mem_realloc(ptr: c_ptr, size: size_t,
+                               description: chpl_mem_descInt_t) : c_void_ptr;
 
-  // String concatenation
-  inline proc +(s: string, x: string) {
-    var cs = __primitive("string_concat", s.c_str(), x.c_str());
-    const ret = toString(cs);
-    // toString steals the c_string_copy, so no need to free it here.
-    return ret;
+  pragma "insert line file info"
+  extern proc chpl_mem_free(ptr: c_ptr): void;
+
+  // TODO: define my own mem descriptors?
+  extern const CHPL_RT_MD_STR_COPY_REMOTE: chpl_mem_descInt_t;
+  extern const CHPL_RT_MD_STR_COPY_DATA: chpl_mem_descInt_t;
+
+  inline proc chpl_string_comm_get(dest: bufferType, src_loc_id: int(64),
+                                   src_addr: bufferType, len: size_t) {
+    __primitive("chpl_comm_get", dest, src_loc_id, src_addr, len);
   }
 
-  inline proc +(s: c_string, x: string) {
-    var cs = __primitive("string_concat", s, x.c_str());
-    const ret = toString(cs);
-    // toString steals the c_string_copy, so no need to free it here.
-    return ret;
+  proc copyRemoteBuffer(src_loc_id: int(64), src_addr: bufferType, len: int): bufferType {
+      const dest = chpl_mem_alloc((len+1).safeCast(size_t), CHPL_RT_MD_STR_COPY_REMOTE): bufferType;
+      chpl_string_comm_get(dest, src_loc_id, src_addr, len.safeCast(size_t));
+      dest[len] = 0;
+      return dest;
   }
 
-  inline proc +(s: string, x: c_string) {
-    var cs = __primitive("string_concat", s.c_str(), x);
-    const ret = toString(cs);
-    // toString steals the c_string_copy, so no need to free it here.
-    return ret;
+  extern proc memmove(destination: c_ptr, const source: c_ptr, num: size_t);
+
+  // pointer arithmetic used for strings
+  inline proc +(a: c_ptr, b: integral) return __primitive("+", a, b);
```

Comment:
I don't think there's anything wrong with supporting + like this for c_ptr. I'd move this to CPtr.chpl. It will be available anyway since String.chpl is imported...